mirrord/agent/src/steal/http/reversible_stream.rs

```diff
@@ -19,25 +22,79 @@ use super::error::HttpTrafficError;
 pub(crate) struct ReversibleStream<const HEADER_SIZE: usize> {
     #[pin]
     stream: TcpStream,
+
     header: [u8; HEADER_SIZE],
+
+    /// We could be writing less then `HEADER_SIZE` bytes into `header`, in two cases:
+    /// 1. Got a timeout before reading as `HEADER_SIZE` bytes.
+    /// 2. Read an EOF after less than `HEADER_SIZE` bytes.
+    /// So we need to always now how many bytes we actually have.
+    header_len: usize,
+
+    /// How many bytes out of the `header` were already read by the owner of self.
     num_forwarded: usize,
 }
 
 impl<const HEADER_SIZE: usize> ReversibleStream<HEADER_SIZE> {
-    pub(crate) async fn read_header(stream: TcpStream) -> Result<Self, HttpTrafficError> {
+    /// Build a Reversible stream from a TcpStream, move on if not done within given timeout.
+    /// Return an Error if there was an error while reading from the TCP Stream.
+    pub(crate) async fn read_header(
+        stream: TcpStream,
+        timeout: Duration,
+    ) -> Result<Self, HttpTrafficError> {
         let mut this = Self {
             stream,
             header: [0; HEADER_SIZE],
+            header_len: 0,
```

Comment:
Add derive(Default) so we don't have to pass all these zeroes at initialization?

Reply:
I can't derive `Default` for `ReversibleStream`, because arrays and `TcpStream`s don't implement `Default`.